Update H4 driver to execute vendor-specific commands for initialization BT Controller before BT Host executes Reset sequence.

npal-cy opened the pull request:
Update H4 driver to execute vendor-specific commands for initialization BT Controller before BT Host executes Reset sequence. Details in https://github.com/zephyrproject-rtos/zephyr/issues/41140.

The following changes was agreed with @carlescufi:
1. add one more field 'setup' in bt_hci_driver type as pointer to HCI vendor-specific setup function
2. add BT_HCI_H4_SETUP Config option in drivers/bluetooth/hci/Kconfig 
3. update hci_init function to call bt_dev.drv->setup() if it not null

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -35,6 +35,17 @@
+ * extansion module if CONFIG_BT_HCI_H4_SETUP is enabled. 
+ */
+#ifdef CONFIG_BT_HCI_H4_SETUP
+extern int h4_setup(void);
```
Considering that you needed the `struct device` pointer in your handler, I wonder if we shouldn't add a layer inside `h4.c` which gives this pointer to the vendor-specific callback. In such a design you'd have a `h4_setup()` function always present in `h4.c` and it would then do something like:
```
#if defined(CONFIG_HCI_H4_SETUP)
        return h4_vnd_setup(h4_dev);
#else
        return 0;
#endif
```

npal-cy replied:
done.

jhedberg commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -131,3 +131,19 @@ config BT_DRIVER_QUIRK_NO_AUTO_DLE
+
+	  The user should generally avoid changing it via menuconfig or in 
+	  configuration files. This option are enabled by the vendor-specific
+	  HCI extension, where the Setup function is implemented.
```
minor: GitHub is pointing out that there's no newline character at the end of this file.

npal-cy replied:
updated.

jhedberg commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -131,3 +131,19 @@ config BT_DRIVER_QUIRK_NO_AUTO_DLE
+config BT_HCI_H4_SETUP
+	bool "Enable HCI vendor-specific Setup function"
+	depends on BT_H4
+	default n
```
n is the "default default", so you can remove this line

npal-cy replied:
updated.

jhedberg reviewed (changes requested):
Looks generally good. A couple of smaller change suggestions in my inline comments.

npal-cy commented:
Hi @jhedberg, thanks for review, i did updates regarding to your comments.

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -35,6 +35,17 @@
+ * h4_vnd_setup function must be implemented in vendor-specific HCI
+ * extansion module if CONFIG_BT_HCI_H4_SETUP is enabled.
+ */
+#ifdef CONFIG_BT_HCI_H4_SETUP
+extern int h4_vnd_setup(const struct device *dev);
+#endif
```
The PR looks pretty good now, and I'm basically ready to approved it, however could you remind me why we wouldn't now want to go for a `__weak` function especially since `h4.c` already uses this tactic for another purpose. @carlescufi do you remember what was the reasoning for a Kconfig instead of `__weak`?

npal-cy replied:
Hi  @jhedberg, 
1. Details about of  propose to use "setup" is in https://github.com/zephyrproject-rtos/zephyr/issues/41140#issuecomment-996090075: 
```
We might want to consider replicating the Linux solution for this, where HCI drivers provide a setup() callback, which would get called some time after the open() callback
```
2. The discussion for implementation details is documented https://discord.com/channels/720317445772017664/733036956392357908/923585062170140715

Hi @carlescufi,
Could you please add your comments here.

Regards,
Nazar

carlescufi replied:
I am pretty sure the reason for this was to save the 4 (or 8) bytes for the `setup` function pointer in the `bt_hci_driver` struct. But right now this is not being conditionally compiled out, so it needs fixing.

carlescufi replied:
This would also be consistent with `conn.h`: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/bluetooth/conn.h#L857

carlescufi replied:
@jhedberg does this make sense to you?

Thalley commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -131,3 +131,18 @@ config BT_DRIVER_QUIRK_NO_AUTO_DLE
 
 	  This has to be enabled when the BLE controller connected is Zephyr
 	  open source controller.
+
+
```
```suggestion

```

Thalley commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -131,3 +131,18 @@ config BT_DRIVER_QUIRK_NO_AUTO_DLE
+	  of the vendor-specific commands sequence to initialize the BT Controller
+	  before the BT Host executes a Reset sequence.
+
+	  The user should generally avoid changing it via menuconfig or in
+	  configuration files. This option are enabled by the vendor-specific
+	  HCI extension, where the Setup function is implemented.
```
Not sure I understand this paragraph. The only way to enable it seem to be from menuconfig or config files.

npal-cy replied:
hi  @Thalley.
user will select BT device  ( e.g BT_CYW43XXX, and it  selects BT_HCI_H4_SETUP)

```
config BT_CYW43XXX
	bool "CYW43XXX BT connectivity"
	select GPIO
	select BT_HCI_H4_SETUP
```
NOTE: support of BT_CYW43XXX will added in another MR.

Thalley replied:
Okay, that explains it. 

Is it OK for the user to `select` manually? If not, then you can remove the text after `bool` to make it a hidden Kconfig, that can only be enabled by e.g. `BT_CYW43XXX`.

npal-cy replied:
For me, it makes sense to hide BT_HCI_H4_SETUP.

Thalley replied:
Then I suggest we do that :)

npal-cy replied:
done.

carlescufi commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -525,11 +536,24 @@ static int h4_open(void)
 	.bus		= BT_HCI_DRIVER_BUS_UART,
 	.open		= h4_open,
 	.send		= h4_send,
+	.setup		= h4_setup
```
#if defined(CONFIG_BT_HCI_H4_SETUP)

npal-cy replied:
after update https://github.com/zephyrproject-rtos/zephyr/pull/41453#discussion_r776332286 , it does not make sense to have #if... here.

now we always have h4_setup:
```
static int h4_setup(void)
{
#if defined(CONFIG_BT_HCI_H4_SETUP)
	/* h4_vnd_setup function must be implemented in vendor-specific HCI
	 * extension module. The option CONFIG_BT_HCI_H4_SETUP should be enabled.
	 */
	return h4_vnd_setup(h4_dev);
#else
	return 0;
#endif
}
```

npal-cy replied:
added #if defined(CONFIG_BT_HCI_H4_SETUP)..

carlescufi commented on include/drivers/bluetooth/hci_driver.h:
```diff
@@ -185,6 +185,16 @@ struct bt_hci_driver {
+	 * BT Controller before BT Host executes Reset sequence.
+	 *
+	 * @return 0 on success or negative error number on failure.
+	 */
```
#if defined(CONFIG_BT_HCI_H4_SETUP)

npal-cy replied:
added

cvinayak commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -525,11 +536,24 @@ static int h4_open(void)
 	return 0;
 }
 
+static int h4_setup(void)
+{
+#if defined(CONFIG_BT_HCI_H4_SETUP)
+	/* h4_vnd_setup function must be implemented in vendor-specific HCI
+	 * extansion module. The option CONFIG_BT_HCI_H4_SETUP should be enabled.
+	 */
+	return h4_vnd_setup(h4_dev);
+#else
+	return 0;
+#endif
+}
+
```
Remove, instead of this hack to supply the `device`, include it in the `bt_hci_driver` so that it is scalable for future drivers that can use the `setup` model and do not have to repeat include such hack of supplying a file static variable to function definitions outside of the driver file.

```suggestion
```

cvinayak commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -525,11 +536,24 @@ static int h4_open(void)
 	.bus		= BT_HCI_DRIVER_BUS_UART,
 	.open		= h4_open,
 	.send		= h4_send,
+	.setup		= h4_setup
```
```suggestion
	.setup = h4_setup,
	.dev = &h4_dev,
```

cvinayak commented on include/drivers/bluetooth/hci_driver.h:
```diff
@@ -185,6 +185,16 @@ struct bt_hci_driver {
+	 *
+	 * @return 0 on success or negative error number on failure.
+	 */
+	int (*setup)(void);
```
```suggestion
	int (*setup)(struct device *dev);
	struct device **dev;
```

cvinayak replied:
```suggestion
	int (*setup)(void *setup_data);
	void *setup_data;
```

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3277,6 +3277,13 @@ static int hci_init(void)
 	int err;
 
+	if (bt_dev.drv->setup) {
+		err = bt_dev.drv->setup();
```
```suggestion
		err = bt_dev.drv->setup(*bt_dev.drv->dev);
```

jhedberg replied:
This doesn't look very clean. So far we haven't exposed any struct device stuff to the host side, and none of the other HCI driver callbacks take this pointer.

jhedberg replied:
Or to put it another way, so far the host hasn't assumed that every HCI driver is built on top of some `struct device` abstraction. E.g. the userchan driver that native_posix uses doesn't have an underlying `struct device` concept (it simply has a file descriptor for read/write data)

cvinayak replied:
`struct device **dev` is a suggestion, `void ** userdata/param` or `struct bt_hci_driver *drv` are other alternative. It is very normal in driver model to have interface provided by a driver be supplied with an associated parent `dev/drv` or `userdata` parameter.

`h4_setup` as glue function to supply `h4_dev` is a hack, will lead to repeated such glue functions in every driver (ex. h5, spi, userchan) file that wish to implement a vendor `setup` interface.

jhedberg replied:
It's certainly not ideal if/when other driver need this kind of functionality, but I was assuming things could be refactored latest then. IIRC the Linux Bluetooth subsystem has a "minidriver" abstraction for vendor-specific setup, which can be reused by multiple HCI drivers.

As for Zephyr, the main reason the HCI driver API never evolved to have `struct device` is that it's a singleton (we don't support multiple controllers the way Linux does). Anyway, I'm not completely opposed to adding a context pointer, which could even be `struct device`, since that's fairly standard in Zephyr, but this needs to be made consistent throughout the HCI driver interface, i.e. it should really be the input to every callback, except `open()`. And then open() would return this pointer, so instead of `int open(void)` you'd have `struct device *open()`. (needs to be checked though what we loose with getting fail/success instead of a specific error code)

cvinayak replied:
>  open() would return this pointer, so instead of int open(void); you'd have struct device *open(void);

I thought about this, `open` to return `struct device *`, it could be ok as return. But we cannot expect `bt_enable` the caller of `open` to store and pass `struct device *dev` to `hci_init` and then thereafter to `setup`, will that be acceptable?

> I'm not completely opposed to adding a context pointer, which could even be struct device, since that's fairly standard in Zephyr, but this needs to be made consistent throughout the HCI driver interface, i.e. it should really be the input to every callback, except open()

Hope we can agree to this, add a `void *` context pointer, it will be generic enough so that drivers can pass data reference (`struct device *` or anything specific to that driver `setup`).

npal-cy replied:
hi @jhedberg, do we have agreement here?

jhedberg replied:
Not really. I was hoping to have time to discuss this with @cvinayak in our weekly Bluetooth call yesterday. One thing that’s noteworthy is that the original proposal would have been 100% backwards compatible without a need to go through the stable API change process. Now this has evolved to something that would immediately break any out-of-tree HCI driver.

cvinayak replied:
@jhedberg Adding `setup`, or adding both `setup` and a `data` now requires the stable API change process? `setup` was added into `struct bt_hci_driver` in the original proposal. With `CONFIG_BT_HCI_H4_SETUP=n`, `struct bt_hci_driver` will remain 100% backward compatible.

jhedberg replied:
Adding setup doesn’t require it, since omitted (not explicitly initialized) struct members are implicitly zeroed (or a NULL pointer in this case). Changing existing callbacks does cause breakage.

cvinayak replied:
This is no change to existing callbacks, the change requested was https://github.com/zephyrproject-rtos/zephyr/pull/41453#discussion_r783080377 or rather this https://github.com/zephyrproject-rtos/zephyr/pull/41453#discussion_r784758440

jhedberg replied:
There are several concerns with that proposal:

 - It’s inconsistent with the other callbacks (why shouldn’t they also get a data pointer?)
 - It makes it impossible to keep the driver struct in ROM if the data is not available at buildtime
 - It’s a bit odd to have to feed data back to the driver that the driver has already itself stored in its own struct
 
So it feels like this proposal is trying to forcefully mould the API to fit the the single use case at hand for a single driver, without looking at the overall driver API design and consistency.

cvinayak replied:
The `struct bt_hci_driver` originally has not followed driver model and was not designed for scalability in terms of adding enhancements; it does not support `struct device *` nor to have a custom parameter to the interfaces in it.

If we are willing to stick to this non-standard way of adding interface function to the driver and then having a glue function as workaround to pass parameters, I have no objections as this driver is already an exception, as long as the PR does not increase flash and RAM for samples that does not need `setup`.

jhedberg replied:
FWIW I’m all in favor of exploring transforming HCI drivers to a pure `struct device` based model, however that’s a different and much larger endeavor than what this PR is trying to address.

jhedberg replied:
@cvinayak I assume that the discussion regarding `struct device` conversion can move to a future PR, however there's one thing I wanted to point out here: even if/when the HCI driver interface is converted to `struct device`, the device pointer given to its callbacks will be that which represents the HCI driver instance itself. In the H4 driver case it will *not* be the device pointer of the UART. Since what's desired in this PR is a H4-specific setup callback (which is able to do UART-specific manipulations, such as getting & setting the baud rate) there would either way have to be a wrapper which passes the UART device pointer rather than the HCI driver device pointer.

cvinayak commented:
I request, a test be added to include regression/coverage in CI, also will serve as an example of intended use of this `setup` interface feature.

npal-cy commented:
hi Guys,

Propose one more approach, which can make hci driver more flexible and easy to update in future with minor changes:

1. instead 'setup' function we will have callback.
```
in hci_driver.h update bt_hci_driver structure with adding callback function:
....
struct bt_hci_driver {
	....
	....
	....
	/** 
	 *  @brief HCI driver callback
	 *
	 * Vendors can use this callback to implement extension for BT HCI 
	 * driver.  
	 * 
	 * @param event - HCI driver Event (refer to bt_hci_driver_event) 
	 * 
	 * @param event_param - Event parameter. It is pointer to void,
	 *                      and should be casted to event-specific type.
	 * 
	 * Refer to bt_hci_driver_event enumeration for full list of 
	 * events and detailed description each of them.
	 *  
	 * @return 0 on success, negative error value on failure
	int (*callback)(bt_hci_driver_event event, void* event_param);
};
....
}
```

2. 'setup' will be as one of possible events in callback.
```
hci_driver.h

/** HCI driver events */
enum bt_hci_driver_event {
	
	/* This event is triggered by BT host on beginning of HCI init. Setup event uses 
	 * to executes vendor-specific commands sequence to initialize BT Controller
	 * before BT Host executes Reset sequence.
	 *
	 * This event does not return Event parameter.
     */
	BT_HCI_EVENT_SETUP                     = 0,
....

};

in hci_core.c
static int hci_init(void)
{
#if defined(CONFIG_BT_HCI_EVENT_SETUP)
	if (bt_dev.drv->callback) {
		/* Sends BT_HCI_EVENT_SETUP event to HCI driver callback to
		 * executes vendor-specific commands sequence to initialize
		 * BT Controller before BT Host executes Reset sequence. */
		err = bt_dev.drv->callback(BT_HCI_EVENT_SETUP, NULL);
		if (err) {
			return err;
		}
	}
#endif /* CONFIG_BT_HCI_EVENT_SETUP */
...
}
```

3. all events will be enabled by Kconfig. For now we have only Setup. But in future it will easy to add something more.
  
4. default HCI callback function will be locate in H4 as __weak (similar to bt_hci_transport_setup )
```
int __weak h4_callback(enum bt_hci_driver_event event, void* event_param)
{
	return 0;
}
```
Vendors can overwrite it in own HCI extension driver:
```
from HCI H4 extension driver (cyw43xxx.c):
...
int h4_callback(enum bt_hci_driver_event event, void* event_param)
{
    int err = 0;
    switch(event)
    {
        case BT_HCI_EVENT_SETUP:
            err = h4_vnd_setup();
            break;
            
        default:
            break;

    }

    return err;
} 
```
5. For HCI extension driver also need access to 'struct device *' of bt_uart. Here we can do the same as H4 do:
```
// get device structure from DT (zephyr_bt_uart)
h4_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_bt_uart));
```
or H4 will provide extra API bt_hci_get_device(), which returns h4_dev from H4.c.

Prototype of above proposal here: https://github.com/npal-cy/zephyr-1/commit/39b3e133bc6cc99f5c837aacb798b56febcf7e0c

npal-cy commented:
I am working on one more enhancement ticket (will open soon). Support low power mode for CYW43xx.
The current version of H4 does not support this. There are two extra lines BT_DEV_WAKE and BT_HOST_WAKE:

BT_DEV_WAKE - Bluetooth device wake-up signal: Signal from the host to the CYW4343X indicating that the host
requires attention.

BT_HOST_WAKE - Host wake-up signal. Signal from the CYW4343X to the host indicating that the CYW4343X requires
attention.

I can handle BT_HOST_WAKE from CY43xxx HCI extension driver (set/release pm_constraint by BT_HOST_WAKE gpio event)
I can't handle BT_DEV_WAKE from CY43xxx HCI extension driver. H4 driver should assert/deassert BT_DEV_WAKE before/after HCI UART TX operation.

The approach with calback, can add this support:
```
>> we add new events (e.g. BT_HCI_EVENT_TRANSFER_DATA_START and BT_HCI_EVENT_TRANSFER_DATA_COMPLETED in hci_driver.h

enum bt_hci_driver_event {
	
	/* This event is triggered by BT host on beginning of HCI initialization. Setup event uses 
	 * to executes vendor-specific commands sequence to initialize BT Controller
	 * before BT Host executes Reset sequence.
	 *
	 * This event does not return Event parameter.
     */
	BT_HCI_EVENT_SETUP                     = 0,

	/* This event indicates start of the HCI data transfer.
	 *
	 * This event does not return Event parameter.
	 */
	BT_HCI_EVENT_TRANSFER_DATA_START       = 1,

	/* This event indicates completion of the HCI data transfer 
	 *
	 * This event does not return Event parameter.
	 */ 
	BT_HCI_EVENT_TRANSFER_DATA_COMPLETED   = 2,
	
};

>> update h4.c to notify BT_HCI_EVENT_TRANSFER_DATA_START and BT_HCI_EVENT_TRANSFER_DATA_COMPLETED events:

static int h4_send(struct net_buf *buf)
{
	BT_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
	int err = 0;

#if defined(CONFIG_BT_HCI_EVENT_TRANSFER_DATA_NOTIFY)
	if (bt_dev.drv->callback) {
		err = bt_dev.drv->callback(BT_HCI_EVENT_TRANSFER_DATA_START, NULL);
		if (err) {
			return err;
		}
	}
#endif /* CONFIG_BT_HCI_EVENT_TRANSFER_DATA_NOTIFY */

	net_buf_put(&tx.fifo, buf);
	uart_irq_tx_enable(h4_dev);

	return err;
}

static inline void process_tx(void)
{
....
....
done:
	tx.type = H4_NONE;
	net_buf_unref(tx.buf);
	tx.buf = net_buf_get(&tx.fifo, K_NO_WAIT);
	if (!tx.buf) {
		uart_irq_tx_disable(h4_dev);
#if defined(CONFIG_BT_HCI_EVENT_TRANSFER_DATA_NOTIFY)
		if (bt_dev.drv->callback) {
			err = bt_dev.drv->callback(BT_HCI_EVENT_TRANSFER_DATA_COMPLETED, NULL);
			if (err) {
				return err;
			}
		}
#endif /* CONFIG_BT_HCI_EVENT_TRANSFER_DATA_NOTIFY */
...
}	
```

npal-cy commented:
> I request, a test be added to include regression/coverage in CI, also will serve as an example of intended use of this `setup` interface feature.

I can not find any test in zephyr\tests\ for existing HCI H4 driver. Please point me, where it locates.

Propose to use CY43xxx hci extension as example of using  `setup` interface feature. Ticket for integration CY43xxx hci extension will be soon.   

This feature was tested on my develop branch (internal) with CY43xxx hci extension driver (logs in attachment). Let me know if it is enough for this MR, i expect existing regression test should covered that we do not break anything if feature is disabled.
[test_hci_feature.log](https://github.com/zephyrproject-rtos/zephyr/files/7897475/test_hci_feature.log)

alwa-nordic commented:
I think this extension should not change the H4 module. The new setup() function will use the HCI interface, right? In that case, the changes should be at all at hci_core level. Logically hci_init() would first create a "raw HCI interface", then hand it to setup(), which results in an "HCI Bluetooth Controller interface". Almost like this PR already does, but the H4 module would not be involved.

This structures the driver stack like this:
VendorSpecificInit <then handover to> Bluetooth Host
HCI
H4
UART

We could consider a driver stack like this:
Bluetooth Host
VendorSpecificInit
HCI
H4
UART

jhedberg commented:
@alwa-nordic the current proposal is simply mirroring the Linux HCI driver API, which has been found to work on a wide variety of HCI drivers and controllers from many different vendors. If you've taken a look at the sample implementation from @npal-cy their vendor setup routine *is* actually very HW, or at least HCI transport dependant, since it's accessing UART APIs as well. So to me it seems this should stay as-is because: 1. it's a proven design from Linux, and 2. what's desired is actually specific to this driver. Note that this is already split into two parts: the HCI driver API change and the change to the H:4 driver to take advantage of the new API.

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -525,11 +525,29 @@ static int h4_open(void)
+ * h4_vnd_setup function must be implemented in vendor-specific HCI
+ * extansion module if CONFIG_BT_HCI_H4_SETUP is enabled.
+ */
+extern int h4_vnd_setup(const struct device *dev);
```
Since this is a public symbol, and in some ways a public API as well, I think it's more appropriate to include it within the Bluetooth namespace, so call this `bt_h4_vnd_setup()`. You could also move it into the scope of the `h4_setup()` function, scene it's not needed or used outside of it.

npal-cy replied:
updated.

jhedberg commented on include/drivers/bluetooth/hci_driver.h:
```diff
@@ -185,6 +185,16 @@ struct bt_hci_driver {
+	 *
+	 * @return 0 on success or negative error number on failure.
+	 */
+	int (*setup)(void);
```
This needs to have `#if defined(BT_HCI_SETUP)` guards now that the Kconfig option applies to the HCI driver interface rather than just the H4 driver. Otherwise this PR is starting to look ready to be merged.

jhedberg reviewed (changes requested):
The content of the PR is fine now, thanks for that! However, the way it's split up is not right anymore, now that the Kconfig option is for the HCI driver interface. The driver interface and host changes should be in one commit and then the H4 driver changes (and nothing else) in a second one. Hopefully you can still quickly do this organisational change to the PR? Thanks.

npal-cy committed:
Bluetooth: host: Update hci_init function to call setup function.

Updated hci_init function to call HCI vendor-specific Setup function
(bt_dev.drv->setup()) on beginning of HCI initialization.

This feature need when the BT Controller requires execution of the
vendor-specific commands sequence to initialize the BT Controller before
the BT Host executes a Reset sequence. To enable this feature the
CONFIG_BT_HCI_SETUP should be enable.

Fixes #41140

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

npal-cy committed:
Bluetooth: hci: added HCI vendor-specific Setup function feature

- Added config BT_HCI_SETUP to enable HCI vendor-specific Setup feature,
- Added pointer to 'setup' function in bt_hci_driver structure.

BT_HCI_SETUP feature is useful when the BT Controller requires execution
of the vendor-specific commands sequence to initialize the BT Controller
before the BT Host executes a Reset sequence. To enable this feature the
CONFIG_BT_HCI_SETUP should be enable.

Fixes #41140

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

npal-cy committed:
Bluetooth: H4: added support for HCI vendor-specific Setup feature.

Updated H4 driver to initialize setup function. Finally bt_h4_vnd_setup
function must be implemented in vendor-specific HCI extension module if
CONFIG_BT_HCI_SETUP is enabled.

BT_HCI_SETUP feature is useful when the BT Controller requires execution
of the vendor-specific commands sequence to initialize the BT Controller
before the BT Host executes a Reset sequence.
To enable this feature the CONFIG_BT_HCI_SETUP should be enable.

Fixes #41140

Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

npal-cy commented:
sorry, one more rebase to fix commit comments length.
https://github.com/zephyrproject-rtos/zephyr/runs/4887017042?check_suite_focus=true

alwa-nordic commented on include/drivers/bluetooth/hci_driver.h:
```diff
@@ -185,6 +185,18 @@ struct bt_hci_driver {
+	 *
+	 * @return 0 on success or negative error number on failure.
+	 */
+#if defined(CONFIG_BT_HCI_SETUP)
```
Should this #if be above the Doxygen comment? I'm not sure how Doxygen will parse this.

alwa-nordic replied:
I just checked. The doc comment needs to be inside the #if. Otherwise the doc comment is applied to the next member (concatenated with that members doc comment) when the #if predicate is false.

alwa-nordic replied:
Also, this member does not appear in the Doxygen output because CONFIG_BT_HCI_SETUP is not defined during doc generation. How do we handle this?

npal-cy replied:
let me check how similar stuff is handled

jhedberg replied:
@npal-cy @alwa-nordic such situations are handled through the `PREDEFINED` list that's defined in `doc/zephyr.doxyfile.in`

jhedberg replied:
However the general preference is to avoid #ifdefs in header files to keep that list as small as possible. E.g. you definitely shouldn't do that for normal API functions that may or may not be available (in fact it prevents using IS_ENABLED() when calling them if the header file uses #ifdef). In this case however, since we decided to avoid the memory overhead when not needed, using the #ifdef in the header file is unavoidable.

alwa-nordic reviewed (changes requested):
The doc comment on bt_hci_driver.setup needs to be inside the #if.
